### ISSUE TITLE
Context Deadline Exceeded errors after upgrade from 1.8.4 to 1.9.4

### ISSUE THREAD

**Issue by zhujik** [cilium/cilium#15237]
**General Information**

- Cilium version (run `cilium version`)
  root@ip-192-168-109-221:/home/cilium# cilium version
  Client: 1.9.4 07b62884c 2021-02-03T11:45:44-08:00 go version go1.15.7 linux/amd64
  Daemon: 1.9.4 07b62884c 2021-02-03T11:45:44-08:00 go version go1.15.7 linux/amd64
- Kernel version (run `uname -a`)
  root@ip-192-168-109-221:/home/cilium# uname -a
  Linux ip-192-168-109-221.eu-central-1.compute.internal 4.14.198-152.320.amzn2.x86_64 #1 SMP Wed Sep 23 23:57:28 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
  EKS 1.18
  Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"windows/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
- Link to relevant artifacts (policies, deployments scripts, ...)  
- Generate and upload a system zip:
  
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```
  
[cilium-sysdump-20210308-105247.zip](https://github.com/cilium/cilium/files/6100467/cilium-sysdump-20210308-105247.zip)

However, https://github.com/cilium/cilium-sysdump/issues/110

---

We upgraded our 1.8.4 installation using the following command:

```
helm repo update
helm upgrade --install cilium \
      cilium/cilium \
      --version 1.9.4 \
      --namespace kube-system \
      --set upgradeCompatibility=1.8 \
      --set eni=true \
      --set ipam.mode=eni \
      --set cni.configMap=cilium-cni-configuration \
      --set egressMasqueradeInterfaces=eth0 \
      --set tunnel=disabled \
      --set nodeinit.enabled=true \
      --set prometheus.enabled=true \
      --set operator.prometheus.enabled=true \
      --set hubble.listenAddress=":4244" \
      --set hubble.relay.enabled=true \
      --set hubble.ui.enabled=true \
      --set hubble.enabled=true \
      --set hubble.metrics.enabled="{dns:query;ignoreAAAA;sourceContext=pod;destinationContext=pod,drop:sourceContext=pod;destinationContext=pod,tcp:sourceContext=pod;destinationContext=pod,flow:sourceContext=pod;destinationContext=pod,port-distribution,icmp,http:sourceContext=pod;destinationContext=pod}"
```

Immediately after the deployment, strange errors occured. We noticed that a lot of pods were not coming past the "ContainerCreating" step, the error was "failed to create pod sandbox, agent failed with error 429 context deadline exceeded"
We see a lot of context deadline exceeded errors in cilium pod logs, such as:
```
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=5 subsys=rate uuid=a2d93315-7fd4-11eb-ad1b-0aebd42dfb00
level=info msg="API call has been processed" error="timed out while waiting to be served with 5 parallel requests: context deadline exceeded" name=endpoint-create processingDuration=0s subsys=rate totalDuration=15.000345293s uuid=a2d93315-7fd4-11eb-ad1b-0aebd42dfb00 waitDurationTotal=0s
```

After a few days, the container creating issue vanished, however the context deadline issues remains. We noticed that some pods do not get processed correctly:

```
> kget pods -n mynamespace -o wide
NAME                                READY   STATUS    RESTARTS   AGE    IP                NODE                                               NOMINATED NODE   READINESS GATES
app-754b6cf6f-q9njv                 1/1     Running   0          3h9m   192.168.86.0      ip-192-168-188-195.eu-central-1.compute.internal   <none>           <none>
app-754b6cf6f-ztz4p                 1/1     Running   0          3h9m   192.168.124.165   ip-192-168-109-221.eu-central-1.compute.internal   <none>           <none>
app2-77894d7fc7-jvcpm               1/1     Running   0          3h9m   192.168.107.62    ip-192-168-109-221.eu-central-1.compute.internal   <none>           <none>
bg-2-5b6f9dbf56-9zzv7               1/1     Running   0          3h9m   192.168.101.31    ip-192-168-109-221.eu-central-1.compute.internal   <none>           <none>
bg-3-59b589569-dqmkm                1/1     Running   0          3h9m   192.168.107.227   ip-192-168-109-221.eu-central-1.compute.internal   <none>           <none>
reportservice-655596bcd6-fwzck      1/1     Running   10         3h9m   192.168.19.135    ip-192-168-109-221.eu-central-1.compute.internal   <none>           <none>
mailtrap-6f448b7477-rdp66           1/1     Running   0          3h9m   192.168.183.47    ip-192-168-188-195.eu-central-1.compute.internal   <none>           <none>
oracledb-exporter-d84b9c4b6-c76wk   1/1     Running   0          3h9m   192.168.67.105    ip-192-168-188-195.eu-central-1.compute.internal   <none>           <none>

> kget cep -n denso
NAME                                ENDPOINT ID   IDENTITY ID   INGRESS ENFORCEMENT   EGRESS ENFORCEMENT   VISIBILITY POLICY   ENDPOINT STATE   IPV4              IPV6
app-754b6cf6f-q9njv                 838           50973                                                                        ready            192.168.86.0
app-754b6cf6f-ztz4p                 348           50973                                                                        ready            192.168.124.165
app2-77894d7fc7-jvcpm               1559          53395                                                                        ready            192.168.107.62
bg-2-5b6f9dbf56-9zzv7               1260          17775                                                                        ready            192.168.101.31
reportservice-655596bcd6-fwzck      760           39084                                                                        ready            192.168.19.135
mailtrap-6f448b7477-rdp66           349           50643                                                                        ready            192.168.183.47
oracledb-exporter-d84b9c4b6-c76wk   3264          36813                                                                        ready            192.168.67.105

> kconnect cilium-5qwvs
root@ip-192-168-109-221:/home/cilium# cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                     IPv6   IPv4              STATUS
           ENFORCEMENT        ENFORCEMENT
348        Disabled           Enabled           50973      k8s:component=app                                                      192.168.124.165   ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=mynamespace
554        Disabled           Disabled          4          reserved:health                                                        192.168.120.24    ready
760        Disabled           Enabled           39084      k8s:component=reportservice                                               192.168.19.135    ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=mynamespace
788        Disabled           Enabled           14335      k8s:component=bg-3                                             192.168.107.227   ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=mynamespace
1260       Disabled           Enabled           17775      k8s:component=bg-2                                                192.168.101.31    ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=mynamespace
1393       Disabled           Disabled          1          k8s:alpha.eksctl.io/cluster-name=itd-dev                                                 ready
                                                           k8s:alpha.eksctl.io/instance-id=i-0ff028c841a551fec
                                                           k8s:alpha.eksctl.io/nodegroup-name=ng-spot-worker-3
                                                           k8s:lifecycle=spot
                                                           k8s:node.kubernetes.io/instance-type=t3a.xlarge
                                                           k8s:role=workers
                                                           k8s:topology.ebs.csi.aws.com/zone=eu-central-1c
                                                           k8s:topology.kubernetes.io/region=eu-central-1
                                                           k8s:topology.kubernetes.io/zone=eu-central-1c
                                                           reserved:host
1559       Disabled           Enabled           53395      k8s:component=app2                                                  192.168.107.62    ready
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=mynamespace
1765       Disabled           Disabled          60640      k8s:io.cilium.k8s.policy.cluster=default                               192.168.10.126    ready
                                                           k8s:io.cilium.k8s.policy.serviceaccount=fluentd-elasticsearch
                                                           k8s:io.kubernetes.pod.namespace=kube-system
                                                           k8s:k8s-app=fluentd-elasticsearch
                                                           k8s:version=v1
2187       Disabled           Enabled           45556      k8s:component=reportservice                                               192.168.124.204   ready
                                                           k8s:io.cilium.k8s.namespace.labels.name=union
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=union
2245       Disabled           Disabled          3945       k8s:app.kubernetes.io/component=query-layer                            192.168.120.231   ready
                                                           k8s:app.kubernetes.io/instance=thanos-query
                                                           k8s:app.kubernetes.io/name=thanos-query
                                                           k8s:app.kubernetes.io/version=v0.18.0
                                                           k8s:io.cilium.k8s.policy.cluster=default
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=prometheus
```

Even though the bg-3 pod seems to be managed by cilium (as indicated in `cilium endpoint list`), the pod is not registered in custom resource `cep`.
In the logs we find something like

```
level=info msg="Releasing key" key="[k8s:component=bg-3 k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=mynamespace]" subsys=allocator
level=info msg="Removed endpoint" containerID= datapathPolicyRevision=0 desiredPolicyRevision=7 endpointID=1211 identity=14335 ipv4= ipv6= k8sPodName=/ subsys=endpoint
```

Is our configuration wrong? Why is cilium not able to process the requests correctly? Please advice

**Comment by zhujik** [comment]
At a quick glance at the other issues I could find https://github.com/cilium/cilium/issues/14403, however, I am not sure if the pod that doesn't show up in the CEP resource has the same root cause or not.

**Comment by zhujik** [comment]
But the difference is that we do not use the AWS CNI in chaining mode, we use Cilium standalone. So I don't think that this is the issue.

**Comment by zhujik** [comment]
This is our Cilium config, if that adds anything:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cilium-cni-configuration
  namespace: kube-system
data:
  cni-config: |
    {
      "cniVersion": "0.3.1",
      "name": "cilium",
      "type": "cilium-cni",
      "enable-debug": false,
      "eni": {
        "first-interface-index": 0,
        "subnet-tags": {
            "cilium-usable-subnet":"true"
          }
      }
    }
```

**Comment by christarazi** [comment]
Hi @zhujik, I looked at the sysdump and I can confirm that there are two missing CEPs that exist as pods:

```
bg-scheduler-59b589569-dqmkm
external-dns-59bd489bb4-g5bxx
```

The first one is mentioned in your post. However, we cannot understand what may have happened for them to go missing without the logs. Do you by chance still have the logs around?

Also, msgs like the following indicate the API rate-limiter kicking in to rate-limit endpoint-creations. That's what a 429 status code means. These msgs are normal if you have a large burst of new pods. Additionally, the rate-limiter can tuned if large bursts are common in your environment. Please see https://docs.cilium.io/en/v1.9/configuration/api-rate-limiting/#id1 for more details. 

```
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=5 subsys=rate uuid=a2d93315-7fd4-11eb-ad1b-0aebd42dfb00
```

Since some time has passed, are you still facing this issue? Have you been able to upgrade to v1.9.5?

**Comment by zhujik** [comment]
@christarazi thank you very much for your response.

In the meantime we pushed the upgrade to 1.9.x back and instead only updated to 1.8.8 without major problems. Unfortunately the logs are lost. The hint to the ratelimit seems suspicous, we rotated the pods in the cluster as is recommended on major cilium upgrades (wasn't there a restart-pods options when doing so), however, this was at most dozens of pods, not hundreds or thousands and we kept 1.9.4 installed on our dev cluster for multiple days without the problems going away. They were even worse, we schedule uptimes of our pods and having a few pods scheduled repeatedly lead to the problems described with pods not being registered and invoking the error.

We will make another attempt to upgrade cilium in the forseeable future and see if the error happens again..
